fix/eventfd: O_NONBLOCK

skuenzer opened the pull request:
### Base target

 - Architecture(s): [all]
 - Platform(s): [all]
 - Application(s): [N/A]


### Additional configuration

 - `CONFIG_LIBPOSIX_EVENT=y`

### Description of changes

Commit fc96850 introduced an complete implementation of handling `FNONBLOCK` and `FASYNC` with `fcntl`, including forwarding of error codes to the caller. Because a standard `EINVAL` handler for `ioctl` was used for `eventfd` file descriptors, setting `O_NONBLOCK` caused a failing assertion.

skuenzer committed:
lib/syscall_shim: uk_prsyscall(): Signed decimal parameter format

Introduces the parameter type `PT_DEC` as format for signed decimal
arguments.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/syscall_shim: uk_prsyscall(): Decode `eventfd`, `eventfd2`

Adds the format definitions for the system calls`eventfd` and `eventfd2`
for decoding with `uk_prsyscall()`.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-event: Debug config option

Introduces a configuration option to enable debug messages for
`lib/posix-event`.

Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-event: `SYS_eventfd2`: Use low-level `fcntl` syscall

The eventfd constructor system call `eventfd2` uses the the `fcntl` system
call to setup the file descriptor flag `O_NONBLOCK` during creation. This
commit makes sure that the low-level implementation is directly called so
that there is no dependency to `errno`.

Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: FUNCTION_ARGUMENTS
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

skuenzer committed:
lib/posix-event: `SYS_eventfd2`: Fix assertion with O_NONBLOCK

Commit
 fc96850 - lib/vfscore: Enable FIONBIO in fcntl(F_SETFL)
introduced a complete implementation of handling `FNONBLOCK` and `FASYNC`
with `fcntl`, including forwarding of error codes to the caller. Because
a standard `EINVAL` handler for `ioctl` was used for `eventfd` file
descriptors, setting `O_NONBLOCK` caused a failing assertion.

Checkpatch-Ignore: USE_NEGATIVE_ERRNO
Signed-off-by: Simon Kuenzer <simon@unikraft.io>

John-Ted reviewed (approved):
Both `eventfd` with `EFD_NONBLOCK` and `fcntl` with `O_NONBLOCK` now work, without any assertion fails.

Reviewed-by: Ioan-Teodor Teugea ioan_teodor.teugea@stud.acs.upb.ro

razvand commented on lib/posix-event/eventfd.c:
```diff
@@ -292,9 +293,25 @@ static int eventfd_vfscore_poll(struct vnode *vnode, unsigned int *revents,
+
+	switch (request) {
+	case FIONBIO:
+		/* eventfd operations look for O_NONBLOCK flag. We just
```
I would expand this to say that `FIONBIO` is the `ioctl` flag corresponding to the `O_NONBLOCK` flag. Otherwise, it seems weird to mention `O_NONBLOCK` with no reference to `FIONBIO`.

skuenzer replied:
👍🏻

flpostolache commented:
@skuenzer Thank you for this PR. There is a typo in the first commit message `an complete implementation`. Also, would it be better if we change the assertion in `eventfd.c` from `UK_ASSERT(ret >= 0)` to `UK_ASSERT(ret == 0);` to be more restrictive? 0 is the only value that can be returned in case of success when fcntl is used with `F_SETFL`. Otherwise, the PR looks good and works on my side.

skuenzer commented:
> @skuenzer Thank you for this PR. There is a typo in the first commit message `an complete implementation`. Also, would it be better if we change the assertion in `eventfd.c` from `UK_ASSERT(ret >= 0)` to `UK_ASSERT(ret == 0);` to be more restrictive? 0 is the only value that can be returned in case of success when fcntl is used with `F_SETFL`. Otherwise, the PR looks good and works on my side.

Sure, I updated it. I was seeing it more relaxed that `fcntl` should just not fail. However, you are right, specification requirs it to return `0` in this case. :+1:
I also fixed the typo that you found. Thanks for finding it.

flpostolache reviewed (approved):
Looks good! @razvand please have another look when you can.
Reviewed-by: Florin Postolache <florin.postolache.of@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
